src/Workspaces/CoreTest/Collections/SegmentedArrayTests.cs

```diff
@@ -0,0 +1,183 @@
+        {
+            var data = new SegmentedArray<IntPtr>(length);
+            Assert.Equal(length, data.Length);
+
+            var items = data.GetTestAccessor().Items;
+            Assert.Equal(length, items.Sum(item => item.Length));
+
+            for (var i = 0; i < items.Length - 1; i++)
+            {
+                Assert.Equal(SegmentedArray<IntPtr>.TestAccessor.SegmentSize, items[i].Length);
+                Assert.True(items[i].Length <= SegmentedArray<IntPtr>.TestAccessor.SegmentSize);
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(TestLengths))]
+        public void TestBasicProperties(int length)
+        {
+            var data = new SegmentedArray<IntPtr>(length);
+
+            Assert.True(data.IsFixedSize);
+            Assert.False(data.IsReadOnly);
+            Assert.False(data.IsSynchronized);
+            Assert.Equal(length, data.Length);
+            Assert.Same(data.GetTestAccessor().Items, data.SyncRoot);
+
+            Assert.Equal(length, ((ICollection)data).Count);
+            Assert.Equal(length, ((ICollection<IntPtr>)data).Count);
+            Assert.Equal(length, ((IReadOnlyCollection<IntPtr>)data).Count);
+        }
+
+        [Theory]
+        [MemberData(nameof(TestLengths))]
+        public void TestIndexer(int length)
+        {
+            var data = new SegmentedArray<IntPtr>(length);
+            ResetToSequence(data);
+
+            for (var i = 0; i < length; i++)
+            {
+                data[i] = (IntPtr)i;
+            }
+
+            for (var i = 0; i < length; i++)
+            {
+                Assert.Equal((IntPtr)i, data[i]);
+            }
+
+            for (var i = 0; i < length; i++)
+            {
+                ref var value = ref data[i];
+                Assert.Equal((IntPtr)i, data[i]);
+                value = IntPtr.Add(value, 1);
+
+                Assert.Equal((IntPtr)(i + 1), value);
+                Assert.Equal((IntPtr)(i + 1), data[i]);
+            }
+
+            ResetToSequence(data);
+            for (var i = 0; i < length; i++)
+            {
+                Assert.Equal((IntPtr)i, ((IReadOnlyList<IntPtr>)data)[i]);
+                data[i] = IntPtr.Add(data[i], 1);
+                Assert.Equal((IntPtr)(i + 1), ((IReadOnlyList<IntPtr>)data)[i]);
+            }
+
+            ResetToSequence(data);
+            for (var i = 0; i < length; i++)
+            {
+                Assert.Equal((IntPtr)i, ((IList<IntPtr>)data)[i]);
+                ((IList<IntPtr>)data)[i] = IntPtr.Add(data[i], 1);
+                Assert.Equal((IntPtr)(i + 1), ((IList<IntPtr>)data)[i]);
+            }
+
+            ResetToSequence(data);
+            for (var i = 0; i < length; i++)
+            {
+                Assert.Equal((IntPtr)i, ((IList)data)[i]);
+                ((IList)data)[i] = IntPtr.Add(data[i], 1);
+                Assert.Equal((IntPtr)(i + 1), ((IList)data)[i]);
+            }
+        }
```

Comment:
do we have tests using a different element size to stress test that?

Reply:
All the tests here use `IntPtr` which varies on 32- and 64-bit runs. I added a `Guid` test as well.